### DISCUSSION TITLE
Why is my tap (C++/CMake) built without optimization flags?

### DISCUSSION THREAD

**Discussion by kimwalisch** [Homebrew/discussions#484] [Tap maintenance and Homebrew development] [answered]
Hi,

I have created a tap (https://github.com/kimwalisch/homebrew-primecount) for one of my C++ projects.

Here is the code of my formula:

```bash
class Primecount < Formula
  desc "Fast prime counting function program and C/C++ library"
  homepage "https://github.com/kimwalisch/primecount"
  url "https://github.com/kimwalisch/primecount/archive/v6.2.tar.gz"
  sha256 "51626a8b2500eb30cba171e1358f3eb589552c8a672babd4eb25e0c9ac054091"
  license "BSD-2-Clause"

  depends_on "cmake" => :build
  depends_on "llvm" => :build
  depends_on "libomp"
  depends_on "primesieve"

  def install
    mkdir "build" do
      # Build primecount using non-default LLVM compiler with libomp (OpenMP)
      system "cmake", "..", "-DCMAKE_CXX_COMPILER=" + Formula["llvm"].bin + "/clang++", "-DBUILD_SHARED_LIBS=ON", "-DBUILD_LIBPRIMESIEVE=OFF", *std_cmake_args
      system "make", "install"
    end
  end

  test do
    system "#{bin}/primecount", "1e10"
  end
end
```

When I install it on macOS Big Sur 11.1 with ```--verbose```, I get:

```bash
brew tap kimwalisch/primecount
brew install primecount --verbose

cmake .. -DCMAKE_CXX_COMPILER=/usr/local/opt/llvm/bin/clang++ -DBUILD_SHARED_LIBS=ON 
-DBUILD_LIBPRIMESIEVE=OFF -DCMAKE_C_FLAGS_RELEASE=-DNDEBUG 
-DCMAKE_CXX_FLAGS_RELEASE=-DNDEBUG 
-DCMAKE_INSTALL_PREFIX=/usr/local/Cellar/primecount/6.2 
-DCMAKE_INSTALL_LIBDIR=lib -DCMAKE_BUILD_TYPE=Release 
-DCMAKE_FIND_FRAMEWORK=LAST -DCMAKE_VERBOSE_MAKEFILE=ON -Wno-dev 
-DCMAKE_OSX_SYSROOT=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk
```

So this includes ```-DCMAKE_BUILD_TYPE=Release```, however the clang++ compilation commands are done without ```-O2``` or ```-O3``` (see below). This issue is present on both macOS x64 and macOS ARM64.

```bash
...
/usr/local/opt/llvm/bin/clang++ -DENABLE_DIV32 -Dlibprimecount_EXPORTS 
-I/tmp/primecount-20210108-27228-q2yczc/primecount-6.2/include 
-isystem /usr/local/include -DNDEBUG -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk 
-fPIC -mpopcnt -fopenmp=libomp -o CMakeFiles/libprimecount.dir/src/RiemannR.cpp.o 
-c /tmp/primecount-20210108-27228-q2yczc/primecount-6.2/src/RiemannR.cpp
...
```

What do I need to change, in order for my tap to be built with optimization flags?

**Comment by SMillerDev** [comment]
I believe Homebrews wrappers filter out `-0` usage.

@Homebrew/brew am I right?

**Comment by kimwalisch** [comment, marked as answer]
> I believe Homebrews wrappers filter out -0 usage.

Thanks for your reply, but I don't fully understand what this means (since I am not a homebrew maintainer/developer). More specifically I don't understand what **Homebrews wrappers** are? How can I change my formula, so that ```-O3``` is not removed?

**Reply by kimwalisch** [reply]
OK, I found it out myself. homebrew does not allow -O2 or -O3, instead homebrew requires to use -Os. Many other people have run into this issue before me and likely many more will run into this issue in the future (because -O3 is the default optimization flag for CMake).

Instead of simply removing -O3 it would be better to replace it with -Os.

**Comment by sjackman** [comment]
Are we sure that `-Os` is the best default for Homebrew? `-O3` is more typical.

**Reply by iMichka** [reply]
We use -02 for linux as this is typical there.
-Os has been used on Mac for years, and just follow the recommendations from the Mac documentation.

Please be aware that -Os does something different with llvm compared to gcc, especially on Mac.

See https://github.com/Homebrew/brew/pull/5516 and https://github.com/Homebrew/brew/pull/4773

**Reply by kimwalisch** [reply]
> We use -02 for linux as this is typical there.
> -Os has been used on Mac for years

What do you think about my suggestion:

Instead of simply removing -O2 and -O3, replace it with homebrew's default optimization flag (-Os on macOS and -O2 on Linux)?

This would significantly simplify writing C/C++ taps for users like myself.

**Reply by iMichka** [reply]
That's what we do. We remove all the -O flags, and add our own flags back.

This happens here for Mac: https://github.com/Homebrew/brew/blob/603a0599cb3d4ba4a42dcc145770b187bad91212/Library/Homebrew/extend/ENV/super.rb#L63
and here for linux: https://github.com/Homebrew/brew/blob/master/Library/Homebrew/extend/os/linux/extend/ENV/super.rb#L15

**Reply by kimwalisch** [reply]
OK, but your current code doesn't work correctly when using a non-default C/C++ compiler. See my initial post, there -O3 was simply removed and my program was compiled without any optimization flags.

**Reply by jonchang** [reply]
> cmake .. -DCMAKE_CXX_COMPILER=/usr/local/opt/llvm/bin/clang++ -DBUILD_SHARED_LIBS=ON 
> /usr/local/opt/llvm/bin/clang++ -DENABLE_DIV32 -Dlibprimecount_EXPORTS 

You're bypassing Homebrew's shims here (it would otherwise use the compiler `/usr/local/Homebrew/Library/Homebrew/shims/super/clang++`) so in any case Homebrew isn't touching your compiler arguments. This suggests a bug in either the formula definition or the cmake file.

If you do want to use Homebrew's shims with brewed llvm (untested):

* set `fails_with` to `:clang` and `:gcc` https://rubydoc.brew.sh/Formula.html#fails_with-class_method
* set `ENV["HOMEBREW_OPTIMIZATION_LEVEL"]` to the desired optimization level

**Reply by carlocab** [reply]
>set `fails_with` to `:clang` and `:gcc`

I've actually tried this before. `brew` will still insist on using `gcc`, and refuse to install if you don't have `gcc` installed. And proceed to use `gcc` to build if you do. I haven't yet figured out why this was happening.

**Comment by kimwalisch** [comment]
This homebrew bug (C/C++ & CMake tap compilation without any optimization flags) has been fixed!

My [tap formula](https://github.com/kimwalisch/homebrew-primecount/blob/v7.2.3/Formula/primecount.rb#L41) now builds with -O3 by default using the non-default LLVM/Clang compiler with OpenMP. Tested on an Apple M1 Mac Mini with the latest macOS 12.0.1 and homebrew. Hence I was unable to remove my workaround.

 Thanks to the homebrew maintainers.

**Reply by ipatch** [reply]
maybe i missed something, but you never clearly stated what the fix was? you said it's working now, and never said what the fix was.

**Comment by kimwalisch** [comment]
When I initially reported this bug in 2021 my formula was compiled without any optimization flags by default (as you can see from my initial bug report). Then about 1 year later I tried it again and my formula was correctly compiled with -O3. I am mainly a homebrew user (who only contributed two formulas for my own projects) I have no idea how or who fixed this bug, but as far as I can tell it is fixed now.

**Reply by ipatch** [reply]
interesting. maybe the superenv for the brew cmd got updated with the `-O3` flag at some point. thanks for the prompt follow up reply 🙏
